add PAR script action (#31188)
DataDog/documentation: content/en/actions/private_actions/run_script.md

| Running privileged commands (`systemctl restart`)   | No    | Because the PAR runs inside a container, it does not have high privilege permissions on the host.                                         |
| Windows tools (PowerShell)                          | No    | Because the PAR runs inside a Linux container, native Windows tools are not supported.                                             |

## Prerequisites

To use the script action, you need:

- **Custom tools**: For CLI tools not included in the base image, you need to create a custom Docker image.
- **PAR Version**: {{< private-action-runner-version "private-action-runner" >}} or later. To create a new PAR, see [Set Up a Private Action Runner][2]. To update your PAR version, see [Update the Private Action Runner][11].

## Set up a PAR script

### Create a script connection

1. After [setting up a PAR][1], navigate to [**Connections**][2].
1. Click **New Connection**.
1. Select **Script**.
1. Enter a **Connection Name**.
1. In the **Private Action Runner** dropdown, select your PAR. 
1. Copy and paste the credential file template into your PAR's configuration directory with real secrets. 

> 1. In **Path to file**, ensure the file path is entered correctly as this field is auto-populated.

Reviewer: We want to encourage people to use the default so that its easier for us to help them debug if needed

```suggestion
1. In **Path to file**, ensure the file path matches the path on your runner's filesystem (the default should be sufficient in most use cases)
```

    command: ["echo", "Hello world"]

  # another script
  echo-parametrized:
    # you can use workflow syntax (https://docs.datadoghq.com/actions/workflows/variables/) to retrieve values from the parameters object
    command: [ "echo", "{{ parameters.echoValue }}" ]
    # you can use JSON schema (https://json-schema.org/) to validate the parameters
    parameterSchema:
      properties:
        echoValue:
          type: string
          const: "world"
      required:
        - echoValue
```

### Using the configured scripts

In your workflow or app, configure the action to use the `runPredefinedScript` with the script name you defined (for example, `echo` or `echo-parametrized`).

**Note**: There are two levels of variable resolution: one at the workflow level and one at the action level inside the runner.

## Advanced usage with custom images

For binaries not available in the base runner image, create a custom image:

```dockerfile
# Dockerfile example
FROM gcr.io/datadoghq/private-action-runner:v1.7.0
USER root
RUN apt update && apt install -y python3
USER dog
```

You can mount complex scripts inside the runner:

```yaml
# docker-compose example
services:
  runner:

>     image: gcr.io/datadoghq/private-action-runner:v1.7.0

Reviewer: ```suggestion
    image: gcr.io/datadoghq/private-action-runner:v1.7.0
    # build: . # if you are using a custom Dockerfile
```

---
title: Run a Script with the Private Action Runner
disable_toc: false
---

## Overview

This page explains how to use the private action runner (PAR), which allows you to run custom scripts and Linux binaries within your Datadog workflows and apps. Unlike standard private actions that call specific APIs or services, the script action gives you the flexibility to execute arbitrary commands, shell scripts, and command-line tools directly from the private action runner in your private network.

<div class="alert alert-warning">

> <strong>Security Notice:</strong> The PAR script action runs within a containerized environment using a dedicated Linux user named `scriptuser` for enhanced security. Datadog enforces container sandboxing and only accepts signed tasks, but you decide which binaries and scripts are allowed. Always review every command you add to the script action allow-list, especially ones that take dynamic user input. Ensure that your actions are configured with the least privileged commands, and carefully review the permissions you share through connections. For more information, see <a href="/actions/connections/?tab=workflowautomation#connection-security-considerations">connection security considerations</a>.

Reviewer 2: ```suggestion
<strong>Security Notice:</strong> The PAR script action runs within a containerized environment using a dedicated Linux user named <code>scriptuser</code> for enhanced security. Datadog enforces container sandboxing and only accepts signed tasks, but you decide which binaries and scripts are allowed. Always review every command you add to the script action allow-list, especially ones that take dynamic user input. Ensure that your actions are configured with the least privileged commands, and carefully review the permissions you share through connections. For more information, see <a href="/actions/connections/?tab=workflowautomation#connection-security-considerations">connection security considerations</a>.
```

## Configuration

Configure script actions through your runner's `config.yaml` file. If you create a new runner and select the script bundle, you get a default configuration.

### Basic configuration

```yaml
# Add the script action to the allowlist
actionsAllowlist:
  - com.datadoghq.script.runPredefinedScript

# Configure different scripts
bundles:
  "com.datadoghq.script":
    runPredefinedScript:
      echo:
        command: ["echo", "Hello world"]
      # Use workflow-like syntax to retrieve values from parameters
      echo-parametrized:
        command: ["echo", "{{ parameters.echoValue }}"]
      echo-nested-parametrized:
        command: ["echo", "{{ parameters.nested.echoValue }}"]
```

### Using the configured scripts

In your workflow or app, configure the action to use the `runPredefinedScript` with the script name you defined (for example, `echo` or `echo-parametrized`).

**Note**: There are two levels of variable resolution: one at the workflow level and one at the action level inside the runner.

## Advanced usage with custom images

For binaries not available in the base runner image, create a custom image:

```dockerfile
# Dockerfile example

> FROM gcr.io/datadoghq/private-action-runner:v1.7.0
> RUN apt update && apt install -y python3

Reviewer: ```suggestion
FROM gcr.io/datadoghq/private-action-runner:v1.7.0
USER root
RUN apt update && apt install -y python3
USER dog
```

Reviewer 3: I thought end user is `scriptuser` ?

```yaml
# docker-compose example
services:
  runner:
    image: gcr.io/datadoghq/private-action-runner:v1.7.0
    volumes:
      - "./config:/etc/dd-action-runner/config"
      - "./scripts:/etc/dd-action-runner-script/scripts:ro"
```

```yaml
# credentials/script.yaml
schemaId: script-credentials-v1
runPredefinedScript:
  python:
    command: ["python3", "/etc/dd-action-runner-script/scripts/script.py"]
  shell:
    command: [ "bash", "/etc/dd-action-runner-script/scripts/script.sh" ]
```

## Updating the PAR

<!-- ADD INFO? -->

To view the latest version of the PAR, check [this URL][6].

Before updating the PAR, you need to have [remote configuration][7] enabled in your organization for the workflow automation mode.

### Docker mode
Navigate to the directory where you started the PAR. Next, navigate to the `config` directory, then the `config.yaml` file. 

Find the current ID of your container:
```bash
docker ps
```

Stop the container:
```bash
docker stop <id>
```

Start a new container with [the latest image][7]. Environment variables are no longer needed. Everything is configured in the `config/config.yaml` file.

Run:
```bash

> docker run -d  -u 0 --platform=linux/x86_64  -p 9016:9016  -v ./config:/etc/dd-action-runner  --health-cmd "curl http://localhost:9016/liveness"  --health-interval 10s  --health-timeout 10s  --health-retries 3 gcr.io/datadoghq/private-action-runner:v1.0.0

Reviewer: yes we recently updated the flag so we should no longer use `--platform` and `--user` maybe we should put the exact command inside the app on the runner's edit page ? This way we can keep them in sync more easily ? But maybe we can start with the doc like this for now

The latest command should be 

```bash
docker run -d \
 --cpus="0.25" \
 --memory="1g"  \
 -e DD_PRIVATE_RUNNER_CONFIG_DIR=/etc/dd-action-runner/config 
 -v ./config:/etc/dd-action-runner/config \
 --health-cmd "curl http://localhost:9016/liveness" \
 --health-interval 10s \
 --health-timeout 10s \
 --health-retries 3 gcr.io/datadoghq/private-action-runner:v1.7.0
```

(essentially what [this](https://app.datadoghq.com/actions/private-action-runners/new) displays but removing all env variables except the config dir one )
